######### The commands used in scalloped spiny lobster *Panulirus homarus homarus* ##########

## 1. Quailty control of illumina reads with fastp (v 0.23.1)

```
fastp -w 50 \
-i FDSW202182427-1r_L2_1.fq.gz \
-I FDSW202182427-1r_L2_2.fq.gz \
-o Pah_illumina_R1.clean.fq.gz \
-O Pah_illumina_R2.clean.fq.gz
```



## 2. Genome survey using kmer distribution generated by the Illumina reads with SOAPec (v 2.01) and GenomeScope (v 2.0)

```
SOAPec -k 17 -s 10 -u 0.2 -t 20 -m 300 -f Pah_illumina_R1_clean.fq Pah_illumina_R2_clean.fq 
Rscript genomescope.R -i out_17merFrq.tsv -k 17 -o kmer17 
```



## 3. Genome assembly

### 3.1 Genome assembly using Wtdbg2 (v2.5)

```
wtdbg2 -x pacbio -k 0 -p 21 -K 1000.049988 -A -S 4.000000 -s 0.050000 -g 0 -X 50.000000 -e 3 -L 0 -o wtdbg2_output \
  --max_depth 50 \
  --node-drop 0.25,0.20 \
  --node-len 1024,1536,2048 \
  --node-max 200 \
  --brute_force 1 \
  P01DY20290441-1_r64054_20201021_074515_2_B01.subreads.bam.fasta.gz \
  P01DY20290441-1_r64054_20201023_020922_1_D01.subreads.bam.fasta.gz \
  P01DY20290441-1_r64116_20201001_080651_1_F02.subreads.bam.fasta.gz
```

### 3.2 Genome assembly using flye (v2.9)

```
flye --pacbio-raw --threads 16 --genome-size 3.13g \
  --out-dir flye_output \
  P01DY20290441-1_r64054_20201021_074515_2_B01.subreads.bam.fasta.gz \
  P01DY20290441-1_r64054_20201023_020922_1_D01.subreads.bam.fasta.gz \
  P01DY20290441-1_r64116_20201001_080651_1_F02.subreads.bam.fasta.gz
```

### 3.3 Polishing of assembly results using Arrow (v8.0) and Pilon (v1.22)

```
## Polishing Wtdbg2 assembly results using Arrow
arrow -o polished_wtdbg2_assembly.fasta wtdbg2_output.fasta \
P01DY20290441-1_r64054_20201021_074515_2_B01.subreads.bam

## Polishing Flye assembly results using Arrow
arrow -o polished_flye_assembly.fasta flye_output/assembly.fasta \
P01DY20290441-1_r64054_20201021_074515_2_B01.subreads.bam

## Merging Arrow polished results using Quickmerge (v0.3)
quickmerge polished_wtdbg2_assembly.fasta polished_flye_assembly.fasta -o merged_polished_assembly.fasta

## Polishing megege assembly results using Arrow
arrow -o polished_assembly_v1.fasta merged_polished_assembly.fasta \
P01DY20290441-1_r64054_20201021_074515_2_B01.subreads.bam

## Polishing the assembly results with the Arrow tool again
arrow -o polished_assembly_v2.fasta polished_assembly_v1.fasta \
P01DY20290441-1_r64054_20201021_074515_2_B01.subreads.bam

## First Pilon polishing using Illumina short read data
pilon --genome polished_assembly_v2.fasta \
  --frags FDSW202182427-1r_L2_1.fq.gz \
  --frags FDSW202182427-1r_L2_2.fq.gz \
  --output pilon_corrected_assembly_v1.fasta \
  --threads 50

## Re-Pilon Polishing using Illumina short read data 
pilon --genome pilon_corrected_assembly_v1.fasta \
  --frags FDSW202182427-1r_L2_1.fq.gz \
  --frags FDSW202182427-1r_L2_2.fq.gz \
  --output draft_assembly.fasta \
  --threads 50
```



## 4. Chromosome-level assembly

### 4.1 Pre-processing of Hi-C data using Juicer (v2.13.06)

```
## Quality control of Hi-C data
fastp -i BDHC200000368-1A_L1.fq.gz -o BDHC200000368-1A_L1.clean.fq.gz -w 8
fastp -i BDHC200000368-1A_L2.fq.gz -o BDHC200000368-1A_L2.clean.fq.gz -w 8
fastp -i BDHC200000368-1A_L3.fq.gz -o BDHC200000368-1A_L3.clean.fq.gz -w 8

juicer_tools pre -d juicer_data --zip \
  BDHC200000368-1A_L1_1.fq.gz BDHC200000368-1A_L1.clean.fq.gz \
  BDHC200000368-1A_L2_1.fq.gz BDHC200000368-1A_L2.clean.fq.gz \
  BDHC200000368-1A_L3_1.fq.gz BDHC200000368-1A_L3.clean.fq.gz \
  -g draft_assembly.fasta \
  -MaxLinkDensity 3 \
  -longest_ 800 \
  -enz mboI \
  -format Sanger \
  -NonInformativeRatio 0 \
  -shortest_ 150 \
  -minREs 50 \
  -fill yes \
  -filter yes \
  -CLUSTER 73
```

### 4.2 Scaffolding of contigs using 3D-DNA pipeline

```
python3 run-3d-dna.py \
  --edit-threshold 0.5 \
  --input juicer_data/merged_output.txt \
  --output-dir 3d_dna_output

mv merged_output.txt P_h_homarus_genome.fasta
```



## 5. Repeat annotation analysis

### 5.1 De novo-predicted repetitive sequence database

```
## Build de novo repeat library
build_lmer_table -sequence P_h_homarus_genome.fasta -freq P_h_homarus_genome.fasta.freq
RepeatScout -sequence P_h_homarus_genome.fasta -freq P_h_homarus_genome.fasta.freq -output P_h_homarus_genome.fasta.out 
cd-hit-est -i P_h_homarus_genome.fasta.out -o P_h_homarus_repeats.fasta -c 0.8 -n 5
cat P_h_homarus_repeats.fasta Repbase.fa > Integrated_repeats.fa 

## Use RepeatModeler (v2.0.1) to predict transposable element families
BuildDatabase -engine ncbi -name P_h_homarus_genome P_h_homarus_genome.fasta
RepeatModeler -database P_h_homarus_genome -output RepeatModeler_output -engine ncbi

## Use Piler (v1.0) to identify transposable elements
Piler -genome P_h_homarus_genome.fasta -output Piler_output

## Use LTR-FINDER (v1.0.6) to predict LTR sequences
LTR-FINDER -C -w 2 -s eukaryotic-tRNAs.fa -O LTRFinder_output P_h_homarus_genome.fasta 
```

### 5.2 Classification of repetitive elements

```
## Use RepeatMasker (v4.1.0) to classify repetitive sequences in the genome
RepeatMasker -a -nolow -no_is -norna -parallel 1 -lib Integrated_repeats.fa -s P_h_homarus_genome.fasta -gff -dir RepeatMasker_output

## Use RepeatProteinMask (v4.1.0) to process protein-coding repetitive sequences
RepeatProteinMask -noLowSimple -pvalue 0.0001 -engine wublast P_h_homarus_genome.fasta

## Use TRF (v4.0.9) to identify tandem repeat sequences
TRF P_h_homarus_genome.fasta 2 7 7 80 10 50 2000 -d -h
```



## 6. Noncoding RNAs prediction

##  6.1 tRNA annotation using tRNAscan-SE (v1.4)

```
tRNAscan-SE -m stat -o tRNA -f structure --thread 20 P_h_homarus_genome.fasta
```

##  6.2 rRNA annotation using BLAST (v2.2.26)

```
blastn -query P_h_homarus_genome.fasta -db rRNA_db -out rRNA_output.txt -evalue 1e-5 -num_threads 50
```

### 6.3 miRNA and snRNA annotation using INFERNAL (v1.0) formRfam (14.5)

```
cmsearch --cpu 20 ./miRNA.DB/Rfam.cm_miRNA P_h_homarus_genome.fasta > genome.fa.miRNA
formatdb -p F -o T -I P_h_homarus_genome.fasta
blastall -p blastn -e 1e-10 -v 10000 -b 10000 -d P_h_homarus_genome.fasta -I animal_rRNA.fa -o genome.fa.miRNA.blast
cmsearch --cpu 20 ./miRNA.DB/Rfam.cm_snRNA P_h_homarus_genome.fasta > genome.fa.snRNA

```



## 7. Gene structure prediction

### 7.1 Gene structure prediction using AUGUSTUS (v3.2.3)

```
augustus --species=fruitfly polished_assembly.fasta > augustus_output.gff
```

### 7.2 Gene structure prediction using GlimmerHMM (v3.02)

```
glimmerhmm P_h_homarus_genome.fasta > glimmerhmm_output.gff
```

### 7.3 Gene structure prediction using SNAP (v2013.11.29)

```
snap gene -sequence P_h_homarus_genome.fasta -output snap_output.gff
```

### 7.4 Gene structure prediction using Geneid (v1.4)

```
geneid P_h_homarus_genome.fasta > geneid_output.gff
```

### 7.5 Gene structure prediction using Genscan (v1.0)

```
genscan P_h_homarus_genome.fasta > genscan_output.gff
```

### 7.6 Homologous comparison using BLAST (v2.2.26)

```
cat D.melanogaster_proteins.fa P.chinensis_proteins.fa E.sinensis_proteins.fa L.vannamei_proteins.fa M.japonicus_proteins.fa P. ornatus_proteins.fa P.trituberculatus_proteins.fa H. americanus_proteins.fa > homology.proteins.fa

blastp -query homology.proteins.fa -subject P_h_homarus_genome.fasta -out blast_output.txt
```

### 7.7 Homologous comparison using Genewise (v2.4.1)

```
genewise P_h_homarus_genome.fasta homology.proteins.fa > genewise_output.gff
```

### 7.8 Combine all prediction genes

```
cat augustus_output.gff glimmerhmm_output.gff snap_output.gff geneid_output.gff genscan_output.gff > P_h_homarus.gff
```

### 7.9 Refinement of transcriptome assembly

```
## Transcriptome data quality control using fastp (v0.23.1)
fastp -w 50 -i BW_G_1.fq.gz -I BW_G_2.fq.gz -o BW_G_1.clean.fq.gz -O BW_G_2.clean.fq.gz
fastp -w 50 -i BW_E_1.fq.gz -I BW_E_2.fq.gz -o BW_E_1.clean.fq.gz -O BW_E_2.clean.fq.gz
fastp -w 50 -i BW_I_1.fq.gz -I BW_I_2.fq.gz -o BW_I_1.clean.fq.gz -O BW_I_2.clean.fq.gz
fastp -w 50 -i BW_H_1.fq.gz -I BW_H_2.fq.gz -o BW_H_1.clean.fq.gz -O BW_H_2.clean.fq.gz
fastp -w 50 -i BW_M_1.fq.gz -I BW_M_2.fq.gz -o BW_M_1.clean.fq.gz -O BW_M_2.clean.fq.gz
fastp -w 50 -i BW_L_1.fq.gz -I BW_L_2.fq.gz -o BW_L_1.clean.fq.gz -O BW_L_2.clean.fq.gz

## De novo assembly using Trinity (v2.11.0)
Trinity --no_version_check --seqType fq --max_memory 400G --CPU 50 \
  --left BW_G_1.clean.fq.gz,BW_E_1.clean.fq.gz,BW_I_1.clean.fq.gz,BW_H_1.clean.fq.gz,BW_M_1.clean.fq.gz,BW_L_1.clean.fq.gz \ 
  --right BW_G_2.clean.fq.gz,BW_E_2.clean.fq.gz,BW_I_2.clean.fq.gz,BW_H_2.clean.fq.gz,BW_M_2.clean.fq.gz,BW_L_2.clean.fq.gz

## Genome guided transcriptome assembly by Hisat (v2.2.1), samtools (v1.12), and Trinity (v2.11.0)
hisat2-build P_h_homarus_genome.fasta P_h_homarus_genome.index -p 20 

hisat2 -p 40 --dta -x P_h_homarus_genome.fasta 
-1 BW_G_1.clean.fq.gz,BW_E_1.clean.fq.gz,BW_I_1.clean.fq.gz,BW_H_1.clean.fq.gz,BW_M_1.clean.fq.gz,BW_L_1.clean.fq.gz \
-2 BW_G_2.clean.fq.gz,BW_E_2.clean.fq.gz,BW_I_2.clean.fq.gz,BW_H_2.clean.fq.gz,BW_M_2.clean.fq.gz,BW_L_2.clean.fq.gz 
-S hisat2_GG.sam

samtools view -bhS hisat2_GG.sam -@ 20 -o hisat2_GG.bam | samtools sort - -@ 20 -m 10G -o hisat2_GG.sorted.bam && samtools index hisat2_GG.sorted.bam 

Trinity --no_version_check \
  --genome_guided_bam hisat2_GG.sorted.bam \
  --max_memory 400G --CPU 100 \
  --genome_guided_max_intron 20000 \
  --output Trinity_GG
```

### 7.10 Identifying open reading frames using PASA (v2.1.0) 
```
cat Trinity.fasta Trinity-GG.fasta > Trinity_all_new.fasta 
seqclean Trinity.fasta \
-v UniVec \
-c 10 seqclean Trinity_all_new.fasta \
-v UniVec \
-c 10 accession_extractor.pl < Trinity.fasta.clean >tdn.accs 

Launch_PASA_pipeline.pl \
-c alignAssembly.config \
-C -R -g Pha.asm.v1.softmask.fa \
-t Trinity_all_new.fasta.clean \
-T -u Trinity_all_new.fasta \
--TDN tdn.accs --ALIGNERS gmap,blat \
--CPU 50 \
--stringent_alignment_overlap 30.0 \
--gene_overlap 50.0 build_comprehensive_transcriptome.dbi \
-c alignAssembly.config \
-t Trinity_all_new.fasta.clean \
--min_per_ID 95 \
--min_per_aligned 30
```



## 8. Functional annotation

### 8.1 Diamond Blast (v0.88.2) the whole NCBI NR database (Release 2021_9_29)
```
diamond makedb --in nr.fa.gz -d nr diamond blastp -e 1e-5 -k 1 -f 6 -p 15 -d nr -q Pah.longest.gene.pep.fasta -o Pah.nr.diamond.result
```

### 8.2 Blast the SWISS-PROT database (Release 2022_01)

```
makeblastdb \
  -in uniprot_sprot.fasta \
  -dbtype prot \
  -out uniprot_sprot.db \
  -parse_seqids 

blastp -query Pah.longest.gene.pep.fasta \
  -db uniprot_sprot.db -out Pah.swissprot.blastp.result \
  -outfmt "6 qseqid sseqid evalue bitscore stitle" \
  -evalue 1e-5 \
  -num_threads 30
```

### 8.3 Protein domain annotation by interproscan (v5.52-86.0) and pfam_scan.pl

```
interproscan.sh \
  -i Pah.longest.gene.pep.fasta \
  -cpu 20 \
  -goterms -iprlookup -pa 

pfam_scan.pl \
  -e_seq 0.05 \
  -e_dom 0.05 \
  -cpu 20 \
  -fasta Pah.longest.gene.pep.fasta \
  -dir Pfam35 \
  -outfile Pah.pfam
```

## 9. Genome assessment

### 9.1 Use BUSCO (v5.8.0) to check the genome completeness

```
busco -c 10 \
  --augustus \
  -m geno \
  --offline \
  --download_path busco_downloads \
  -i P_h_homarus_genome.fasta \
  -l arthropoda_odb12 \
  -o busco_augustus
```

## 9.2 Detect gene orthologs using tblastn, Augustus, and HMMER

```
tblastn -query protein_query.fasta -db P_h_homarus_genome.fasta -evalue 1e-5 -outfmt 6 -out tblastn_results.txt 

augustus --species=species_model --genome=P_h_homarus_genome.fasta --gff3=on > augustus_predictions.gff 

hmmer3 -o hmmer_results.txt \
  --tblout hmmer_tblout.txt \
  --cpu 4 --evalue 1e-5 \
  --domtblout hmmer_domtblout.txt protein_domain_model.hmm P_h_homarus_genome.fasta
```

### 9.3 Use Merqury (https://github.com/marbl/merqury) to check the QV and genome completeness

```
meryl k=21 memory=500G count output reads.meryl ../BDHC200000368_clean_R1.fq.gz ../BDHC200000368_clean_R2.fq.gz 

merqury.sh reads.meryl/ ../P_h_homarus_genome.fasta merqury_output
```

### 9.4 Map paired-end illumina sequencing DNA reads to genome with bwa (v0.7.8)

```
bwa index Pah.asm.v1.fa bwa mem -t 10 -R '@RG\tID:Pah_illumina\tSM:Pah_illumina\tPL:illumina' ../P_h_homarus_genome.fasta Pah_illumina_R1_clean.fq.gz Pah_illumina_R2_clean.fq.gz 2>Pah_illumina.bwa.log 

samtools sort -@ 10 -m 10G -o Pah_illumina.sort.bam 

samtools flagstat -@ 10 Pah_illumina.sort.bam > Pah.bam.flagstat
```
### 9.5 Genome-wide homology analysis between  P. h. homarus and P. ornatus used MCScanX within the JCVI toolkit (v1.1.12)

```
python -m jcvi.formats.gff bed --type=gene --key=ID  pah.longest.gff > pah.bed
python -m jcvi.formats.gff bed --type=gene --key=ID poh.longest.gff > poh.bed


python -m jcvi.formats.gff bed --type=gene --key=ID ../pah.longest.gff3 >  pah.bed
python -m jcvi.formats.gff bed --type=gene --key=ID ../poh.longest.gff3 > poh.bed

sed 's/>hapA.*gene=/>/g'  pah.pep.fa | awk '{print $1}' >  pah.pep
sed 's/>hapB.*gene=/>/g'  poh.pep.fa | awk '{print $1}' > poh.pep

python3 -m jcvi.compara.catalog ortholog pah pao --no_strip_names
python3 -m jcvi.compara.synteny screen --minspan=15 --minsize=5 --simple  pah.poh.anchors  pah.poh.anchors.new
python simple2links.py  pah.poh.anchors.simple

